src/compiler/crystal/semantic/call_error.cr

```diff
@@ -194,6 +176,311 @@ class Crystal::Call
     raise message, owner_trace
   end
 
+  private def check_block_mismatch(call_errors, owner, def_name)
+    return unless call_errors.all?(BlockMismatch)
+
+    error_message =
+      if block
+        "'#{full_name(owner, def_name)}' is not expected to be invoked with a block, but a block was given"
+      else
+        "'#{full_name(owner, def_name)}' is expected to be invoked with a block, but no block was given"
+      end
+
+    raise error_message
+  end
+
+  private def check_missing_named_arguments(call_errors, owner, defs, arg_types, inner_exception)
+    return unless call_errors.all?(MissingNamedArguments)
+
+    call_errors = call_errors.map &.as(MissingNamedArguments)
+    all_missing_names = call_errors.flat_map(&.names).uniq
+    missing_names_in_all_overloads = all_missing_names.select do |missing_name|
+      call_errors.all? &.names.includes?(missing_name)
+    end
+    return if missing_names_in_all_overloads.empty?
+
+    error_message = String.build do |str|
+      if missing_names_in_all_overloads.size == 1
+        str << "missing argument: #{missing_names_in_all_overloads.first}"
+      else
+        str << "missing arguments: #{missing_names_in_all_overloads.join ", "}"
+      end
+      str.puts
+      str.puts
+      str << "Overloads are:"
+      append_matches(defs, arg_types, str)
```

Comment:
This part is repeated in most of the methods as far as I can tell, it should be identical every time. I think it would fit great in a helper method.

Reply:
I'll do it

Reply:
Done! Only one of them remains different now, but because it's different (it's the most complex case, the one about argument type mismatch)

Reply:
Done!